src/opts/opts_lgrngn.hpp

```diff
@@ -111,21 +111,83 @@ void setopts_micro(
-        case_ptr->mean_rd1, // parameters
-        case_ptr->mean_rd2,
-        case_ptr->sdev_rd1,
-        case_ptr->sdev_rd2,
-        case_ptr->n1_stp,
-        case_ptr->n2_stp
-      )
-    );
 
+  {
+    if(user_params.n1_stp * si::cubic_metres >= 0 && user_params.n2_stp * si::cubic_metres >= 0) {
+      if(user_params.kappa1 == user_params.kappa2){
+        rt_params.cloudph_opts_init.dry_distros.emplace(
+          user_params.kappa1,
+          std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+            user_params.mean_rd1,
+            user_params.mean_rd2,
+            user_params.sdev_rd1,
+            user_params.sdev_rd2,
+            user_params.n1_stp,
+            user_params.n2_stp
+          )
+        );
+      } else {
+        rt_params.cloudph_opts_init.dry_distros.emplace(
+          user_params.kappa1,
+          std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+            user_params.mean_rd1,
+            thrust_real_t(1.0e-6) * si::metres,
+            user_params.sdev_rd1,
+            thrust_real_t(1.2),
+            user_params.n1_stp,
+            thrust_real_t(0) / si::cubic_metres
+          )
+        );
+        rt_params.cloudph_opts_init.dry_distros.emplace(
+          user_params.kappa2,
+          std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+            thrust_real_t(1.0e-6) * si::metres,
+            user_params.mean_rd2,
+            thrust_real_t(1.2),
+            user_params.sdev_rd2,
+            thrust_real_t(0) / si::cubic_metres,
+            user_params.n2_stp
+          )
+        );
+      }      
+    } else if(user_params.n1_stp * si::cubic_metres >= 0 && user_params.n2_stp * si::cubic_metres < 0) {
+      rt_params.cloudph_opts_init.dry_distros.emplace(
+        user_params.kappa1,
+        std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+          user_params.mean_rd1,
+          thrust_real_t(1.0e-6) * si::metres,
+          user_params.sdev_rd1,
+          thrust_real_t(1.2),
+          user_params.n1_stp,
+          thrust_real_t(0) / si::cubic_metres
+        )
+      );
+    } else if(user_params.n1_stp * si::cubic_metres < 0 && user_params.n2_stp * si::cubic_metres >= 0) {
+      rt_params.cloudph_opts_init.dry_distros.emplace(
+        user_params.kappa2,
+        std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+          thrust_real_t(1.0e-6) * si::metres,
+          user_params.mean_rd2,
+          thrust_real_t(1.2),
+          user_params.sdev_rd2,
+          thrust_real_t(0) / si::cubic_metres,
+          user_params.n2_stp
+        )
+      );
+    } else {
+      rt_params.cloudph_opts_init.dry_distros.emplace(
+        case_ptr->kappa,
+        std::make_shared<setup::log_dry_radii<thrust_real_t>> (
+          case_ptr->mean_rd1,
+          case_ptr->mean_rd2,
+          case_ptr->sdev_rd1,
+          case_ptr->sdev_rd2,
+          case_ptr->n1_stp,
+          case_ptr->n2_stp
+        )
+      );
+    }
+ 
```

Comment:
I think that this if-else structure could be simplified:
1. if user_params.n1>=0 - emplace mode 1
2. if user_params.n2>=0 - emplace mode 2
3. if both user_params.n1 and n2 < 0 - emplace default case modes

same goes for opts_blk_2m

Reply:
This would work for `opts_blk_2m`, but I don't think it will work here because of the caveat that when `kappa1 = kappa2` you can't emplace the modes separately.

Reply:
It would be nice to throw a runtime_error if 
`(user_params.n1>=0 && user_params.n2>=0) && user_params.kappa1==user_params.kappa2`

Then, I think that the if-else structure I proposed previously would work well.

Reply:
I don't want that to throw a runtime error though. The user should be able to set two different size modes (with different mean_rd) but the same kappa. For example, if you want to run a dycoms simulation but where the larger mode is even larger (e.g. `mean_rd2 = 80nm`), but both modes have `kappa=0.61`.

Reply:
The function pointed by the value of the dry_distros map in principle can have any number of modes.
This is why I don't think that there is need for multiple entries with the same kappa.
I suppose that having any number of modes could be implemented with variadic arguments (and maybe variadic templates).

Before this is implemented, I propose we do the following:
1. throw runtime_error if kappa1==kappa2 ONLY in opts_lgrngn (kappa1==kappa2 does not work with lgrngn anyway)
2. have the same if-else structure for setting dry_distros as in opts_blk_2m